### PR TITLE
Bug 1903206: Check Ready condition status in Endpointslices

### PR DESCRIPTION
**Pull request by frobware**
When aggregating endpointslices check the Ready condition status. If
it is not ready then accumulate those addresses into the
NotReadyAddresses bucket. Although the router doesn't actually
use NotReadyAddresses we set this to have a faithful and technically
correct translation of EndpointSlices to Endpoints.

EndpointSlices were introduced in 4.6: see #154 

Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1903206

### CODE REVIEW

**Comment by frobware** [comment]
```
[sig-apps] ReplicaSet should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-apps] ReplicationController should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-network] pods should successfully create sandboxes by getting pod
[sig-network] pods should successfully create sandboxes by reading container

```

/retest

File: pkg/router/controller/router_controller.go [hunk 1]
```diff
@@ -300,7 +301,12 @@ func convertEndpointSliceToEndpointSubset(items []discoveryv1beta1.EndpointSlice
 				if items[i].Endpoints[j].Hostname != nil {
 					epa.Hostname = *items[i].Endpoints[j].Hostname
 				}
-				addresses = append(addresses, epa)
+				// A nil Ready condition indicates an unknown state and should be interpreted as ready.
```
Comment (sgreene570) [inline pkg/router/controller/router_controller.go:304]:
```suggestion
				// A nil Ready condition indicates an unknown state and should be interpreted as not ready.
```
Right?

Comment (sgreene570) [inline pkg/router/controller/router_controller.go:304]: I'm confused how a `nil` Ready condition would mean we should interpret the endpoint as ready. Are there docs, or anything of that sorts, on this?

Comment (sgreene570) [reply pkg/router/controller/router_controller.go:304]: Ah, turns out there is https://github.com/kubernetes/api/blob/master/discovery/v1beta1/types.go#L116-L123

Comment (sgreene570) [reply pkg/router/controller/router_controller.go:304]:
```
// EndpointConditions represents the current condition of an endpoint.
type EndpointConditions struct {
	// ready indicates that this endpoint is prepared to receive traffic,
	// according to whatever system is managing the endpoint. A nil value
	// indicates an unknown state. In most cases consumers should interpret this
	// unknown state as ready. For compatibility reasons, ready should never be
	// "true" for terminating endpoints.
	// +optional
	Ready *bool `json:"ready,omitempty" protobuf:"bytes,1,name=ready"`
```

Comment (frobware) [reply pkg/router/controller/router_controller.go:304]: > Ah, turns out there is https://github.com/kubernetes/api/blob/master/discovery/v1beta1/types.go#L116-L123

This is why I added the comment - just some additional clarity/context if you don't look that the docs and just read the code as it is now.

File: pkg/router/controller/router_controller.go [hunk 2]
```diff
@@ -321,11 +327,13 @@ func convertEndpointSliceToEndpointSubset(items []discoveryv1beta1.EndpointSlice
...
 
 		endpointsubset.SortAddresses(addresses, addressOrderByFuncs)
+		endpointsubset.SortAddresses(notReadyAddresses, addressOrderByFuncs)
 		endpointsubset.SortPorts(ports, portOrderByFuncs)
 
 		subsets = append(subsets, kapi.EndpointSubset{
-			Addresses: addresses,
-			Ports:     ports,
+			Addresses:         addresses,
+			NotReadyAddresses: notReadyAddresses,
```
Comment (Miciah) [inline pkg/router/controller/router_controller.go:335]: It doesn't look like the router actually uses `NotReadyAddresses`...  But we can do this to have a faithful and technically correct translation of EndpointSlices to Endpoints.

**Comment by frobware** [comment]
73 unit test failures.

/test e2e-agnostic

File: pkg/router/controller/router_controller.go (hunk 2 above)
Comment (frobware) [reply pkg/router/controller/router_controller.go:335]: Added this comment to the PR description. Thanks!

**Comment by frobware** [comment]
```
 Using must-gather plug-in image: registry.redhat.io/openshift4/ose-must-gather:latest
Unable to connect to the server: dial tcp 35.237.165.255:6443: i/o timeout
```
/retest

**Comment by frobware** [comment]
```
[sig-apps] ReplicationController should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s]
[sig-apps] ReplicaSet should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s]
```
/retest

**Comment by frobware** [comment]
```
[sig-apps] ReplicaSet should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s]: [sig-apps] ReplicaSet should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s] 
[sig-apps] ReplicationController should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s] 
[sig-network] pods should successfully create sandboxes by writing network status 	0s
[sig-network] pods should successfully create sandboxes by getting pod 	0s
[sig-network] pods should successfully create sandboxes by reading container 	0s
```
/retest

**Comment by frobware** [comment]
```
[sig-apps] ReplicationController should serve a basic image on each replica with a private image [Suite:openshift/conformance/parallel] [Suite:k8s] 

```

- ReplicationController failure is this bug: https://bugzilla.redhat.com/show_bug.cgi?id=1903674
- And there's a PR to disable the test: https://github.com/openshift/origin/pull/25732

**Comment by frobware** [comment]
Re: 

```
[sig-network] pods should successfully create sandboxes by writing network status 	0s
[sig-network] pods should successfully create sandboxes by getting pod 	0s
[sig-network] pods should successfully create sandboxes by reading container
```

The pod unit test failures look like they are covered by this bug: https://bugzilla.redhat.com/show_bug.cgi?id=1882750#c29

**Pull request merged**